File: subly_ios/Subly/Subly/Data/Local/Repositories/CategoryRepositoryImpl.swift
import Foundation
import SwiftData
import Combine

enum CategoryError: LocalizedError {
    case inUse(count: Int)

    var errorDescription: String? {
        switch self {
        case .inUse(let count):
            return "Cannot delete category: used by \(count) subscription\(count == 1 ? "" : "s")."
        }
    }
}

final class CategoryRepositoryImpl: CategoryRepository {
    private let modelContext: ModelContext
    private let subscriptionRepository: SubscriptionRepository
    private let subject = PassthroughSubject<Void, Never>()

    init(modelContext: ModelContext, subscriptionRepository: SubscriptionRepository) {
        self.modelContext = modelContext
        self.subscriptionRepository = subscriptionRepository
    }

    // MARK: - Reactive reads

    func getAllCategories() -> AnyPublisher<[Category], Never> {
        subject
            .prepend(())
            .map { [weak self] _ in self?.fetchAll() ?? [] }
            .eraseToAnyPublisher()
    }

    func getAllCategoriesWithCount() -> AnyPublisher<[CategoryWithCount], Never> {
        Publishers.CombineLatest(getAllCategories(), subscriptionRepository.getAllSubscriptions())
            .map { categories, subscriptions in
                let counts = Dictionary(grouping: subscriptions, by: { $0.categoryId })
                    .mapValues { $0.count }
                return categories.map { CategoryWithCount(category: $0, subscriptionCount: counts[$0.id] ?? 0) }
            }
            .eraseToAnyPublisher()
    }

    func getCategoryById(_ id: UUID) -> AnyPublisher<Category?, Never> {
        subject
            .prepend(())
            .map { [weak self] _ in self?.fetchAll().first(where: { $0.id == id }) }
            .eraseToAnyPublisher()
    }

    func getUsageCount(for id: UUID) async -> Int {
        let descriptor = FetchDescriptor<SubscriptionModel>(
            predicate: #Predicate { $0.categoryId == id }
        )
        return (try? modelContext.fetch(descriptor).count) ?? 0
    }

    // MARK: - Mutations

    func insertCategory(_ category: Category) async throws {
        let model = CategoryModel(from: category)
        modelContext.insert(model)
        try modelContext.save()
        subject.send()
    }

    func updateCategory(_ category: Category) async throws {
        let descriptor = FetchDescriptor<CategoryModel>(
            predicate: #Predicate { $0.id == category.id }
        )
        if let existing = try modelContext.fetch(descriptor).first {
            existing.update(from: category)
            try modelContext.save()
            subject.send()
        }
    }

    func deleteCategory(_ category: Category) async throws {
        let count = await getUsageCount(for: category.id)
        guard count == 0 else { throw CategoryError.inUse(count: count) }

        let descriptor = FetchDescriptor<CategoryModel>(
            predicate: #Predicate { $0.id == category.id }
        )
        if let existing = try modelContext.fetch(descriptor).first {
            modelContext.delete(existing)
            try modelContext.save()
            subject.send()
        }
    }

    // MARK: - Private

    private func fetchAll() -> [Category] {
        let descriptor = FetchDescriptor<CategoryModel>(sortBy: [SortDescriptor(\.displayName)])
        return (try? modelContext.fetch(descriptor))?.map { $0.toDomain() } ?? []
    }
}

File: subly_android/app/src/main/java/net/cynreub/subly/data/repository/CategoryRepositoryImpl.kt
package net.cynreub.subly.data.repository

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import net.cynreub.subly.data.local.dao.CategoryDao
import net.cynreub.subly.data.mapper.toCategoryWithCount
import net.cynreub.subly.data.mapper.toDomain
import net.cynreub.subly.data.mapper.toEntity
import net.cynreub.subly.domain.model.Category
import net.cynreub.subly.domain.repository.CategoryRepository
import net.cynreub.subly.domain.sync.SyncProvider
import net.cynreub.subly.ui.categories.CategoryWithCount
import java.util.UUID
import javax.inject.Inject

class CategoryRepositoryImpl @Inject constructor(
    private val categoryDao: CategoryDao,
    private val syncProvider: SyncProvider
) : CategoryRepository {

    override fun getAllCategories(): Flow<List<Category>> =
        categoryDao.getAllCategories().map { it.map { entity -> entity.toDomain() } }

    override fun getAllCategoriesWithCount(): Flow<List<CategoryWithCount>> =
        categoryDao.getAllCategoriesWithCount().map { it.map { entity -> entity.toCategoryWithCount() } }

    override fun getCategoryById(id: UUID): Flow<Category?> =
        categoryDao.getCategoryById(id.toString()).map { it?.toDomain() }

    override suspend fun getUsageCount(id: UUID): Int =
        categoryDao.getUsageCount(id.toString())

    override suspend fun insertCategory(category: Category) {
        categoryDao.insertCategory(category.toEntity())
        syncProvider.upsertCategory(category)
    }

    override suspend fun updateCategory(category: Category) {
        categoryDao.updateCategory(category.toEntity())
        syncProvider.upsertCategory(category)
    }

    override suspend fun deleteCategory(category: Category) {
        val count = categoryDao.getUsageCount(category.id.toString())
        if (count > 0) {
            throw IllegalStateException(
                "Cannot delete \"${category.displayName}\" — it is used by $count subscription(s). Reassign them first."
            )
        }
        categoryDao.deleteCategory(category.toEntity())
        syncProvider.deleteCategory(category.id)
    }
}
